providers/implementations/signature/rsa.c

```diff
@@ -0,0 +1,884 @@
+            if (!setup_tbuf(prsactx))
+                return 0;
+            if (!RSA_padding_add_PKCS1_PSS_mgf1(prsactx->rsa,
+                                                prsactx->tbuf, tbs,
+                                                prsactx->md, prsactx->mgf1_md,
+                                                prsactx->saltlen)) {
+                ERR_raise(ERR_LIB_PROV, ERR_LIB_RSA);
+                return 0;
+            }
+            ret = RSA_private_encrypt(RSA_size(prsactx->rsa), prsactx->tbuf,
+                                      sig, prsactx->rsa, RSA_NO_PADDING);
+        } else {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_PADDING_MODE);
+            return 0;
+        }
+    } else {
+        ret = RSA_private_encrypt(tbslen, tbs, sig, prsactx->rsa,
+                                  prsactx->pad_mode);
+    }
+
+    if (ret <= 0) {
+        ERR_raise(ERR_LIB_PROV, ERR_LIB_RSA);
+        return 0;
+    }
+
+    *siglen = ret;
+    return 1;
+}
+
+static int rsa_verify_recover(void *vprsactx,
+                              unsigned char *rout,
+                              size_t *routlen,
+                              size_t routsize,
+                              const unsigned char *sig,
+                              size_t siglen)
+{
+    PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
+    int ret;
+
+    if (rout == NULL) {
+        *routlen = RSA_size(prsactx->rsa);
+        return 1;
+    }
+
+    if (!setup_md(prsactx))
+        return 0;
+
+    if (prsactx->md != NULL) {
+        if (prsactx->pad_mode == RSA_X931_PADDING) {
+            if (!setup_tbuf(prsactx))
+                return 0;
+            ret = RSA_public_decrypt(siglen, sig, prsactx->tbuf, prsactx->rsa,
+                                     RSA_X931_PADDING);
+            if (ret < 1) {
+                ERR_raise(ERR_LIB_PROV, ERR_LIB_RSA);
+                return 0;
+            }
+            ret--;
+            if (prsactx->tbuf[ret] != RSA_X931_hash_id(EVP_MD_type(prsactx->md))) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_ALGORITHM_MISMATCH);
+                return 0;
+            }
+            if (ret != EVP_MD_size(prsactx->md)) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST_LENGTH);
+                return 0;
+            }
+            if (rout && (routsize == 0 || routsize >= (size_t)ret))
+                memcpy(rout, prsactx->tbuf, ret);
+        } else if (prsactx->pad_mode == RSA_PKCS1_PADDING) {
+            size_t sltmp;
+
+            ret = int_rsa_verify(EVP_MD_type(prsactx->md),
+                                 NULL, 0, rout, &sltmp,
+                                 sig, siglen, prsactx->rsa);
+            if (ret <= 0) {
+                ERR_raise(ERR_LIB_PROV, ERR_LIB_RSA);
+                return 0;
+            }
+            ret = sltmp;
+        } else {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_PADDING_MODE);
+            return 0;
+        }
+    } else {
```

Comment:
Should we be verifying the pad mode in this event? i.e. that its not one of the ones handled above

Reply:
Why?  Doesn't `RSA_public_decrypt()` do that?

Reply:
Ah yes! You're right. It does.